desktop-widgets/importgps.cpp

```diff
@@ -0,0 +1,294 @@
+// SPDX-License-Identifier: GPL-2.0
+#include "desktop-widgets/importgps.h"
+
+/* Import dive coordinates from a GPS device and synchronise them with the dive profile information
+   of a dive computer. This file contains the infrastructure to:
+   1) Read a .GPX file from a GPS system.
+   2) Find the first gpx trackpoint that follows after the start of the dive.
+   3) Allow modification of the coordinates dependent on international time zone and
+	      on differences in local time settings between the GPS and the dive computer.
+   4) Saving the coordinates into the Coordinates text box in the Dive Site Edit panel
+	      and which which causes the map to show the location of the dive site.
+   The structure coords is used to store critical information.  */
+ImportGPS::ImportGPS(QWidget *parent, QString fileName, class Ui::LocationInformation *LocationUI) : QDialog(parent),
+	fileName(fileName), LocationUI(LocationUI)
+{
+	ui.setupUi(this);
+	connect(ui.timeDiffEdit, SIGNAL(timeChanged(const QTime)), this, SLOT(timeDiffEditChanged()));
+	connect(ui.timeZoneEdit, SIGNAL(timeChanged(const QTime)), this, SLOT(timeZoneEditChanged()));
+	connect(ui.timezone_backwards, SIGNAL(toggled(bool)), this, SLOT(changeZoneBackwards()));
+	connect(ui.timezone_forward, SIGNAL(toggled(bool)), this, SLOT(changeZoneForward()));
+	connect(ui.diff_backwards, SIGNAL(toggled(bool)), this, SLOT(changeDiffBackwards()));
+	connect(ui.diff_forward, SIGNAL(toggled(bool)), this, SLOT(changeDiffForward()));
+	connect(ui.GPSbuttonBox, SIGNAL(clicked(QAbstractButton *)), this, SLOT(buttonClicked(QAbstractButton *)));
+	coords.settingsDiff_offset = 0;
+	coords.timeZone_offset = 0;
+	coords.lat = 0;
+	coords.lon = 0;
+	pixmapSize = (int) (ui.diveDateLabel->height() / 2);
+}
+
+void ImportGPS::buttonClicked(QAbstractButton *button)
+{
+	if (ui.GPSbuttonBox->buttonRole(button) == QDialogButtonBox::AcceptRole) {
+		// Write the coordinates in decimal degree format to the Coordinates QLineEdit of the LocationaInformationWidget UI:
+		LocationUI->diveSiteCoordinates->setText(QString::number(coords.lat) + ", " + QString::number(coords.lon));
+		LocationUI->diveSiteCoordinates->editingFinished();
+	} else {
+		close();
+	}
+}
+
+//  Read text from the present position in the file until
+//  the first 'delim' character is encountered.
+int ImportGPS::getSubstring(QFile *file, QString *bufptr, char delim)
+{
+	char c;
+	bufptr->clear();
+	do {
+		if (file->read(&c, 1) <= 0) // EOF
+			return 1;
+		if ( c == delim) break;
+		bufptr->append(QChar(c));
+	} while (c != delim);
+	return 0;
+}
+
+// Find the next occurence of a specified target GPX element in the file,
+// characerised by a "<xxx " or "<xxx>" character sequence.
+// 'target' specifies the name of the element searched for.
+// termc is the ending character of the element name search = '>' or ' '.
+int ImportGPS::findXmlElement(QFile *fileptr, QString target, QString *bufptr, char termc)
```

Comment:
a nice to have would be to have unit tests for the XML parsing logic here.
but as someone commented here https://github.com/Subsurface-divelog/subsurface/pull/2469#discussion_r368404123
i think this logic can be replaced with Qt's XML objects.

Reply:
I agree. But let's just get this code into master first to avoid the need for another rebase. Then we can add the trimmings. There is also some work to do with respect to the default directory with gpx files.